File: HyowonSin/season2/_230809/PGS_뉴스클러스터링.py
def solution(str1, str2):
    str1 = str1.lower()
    str2 = str2.lower()

    setStr1 = []
    setStr2 = []

    for i in range(len(str1)-1):
        if str1[i].isalpha() and str1[i+1].isalpha():
            setStr1.append(str1[i:i+2])

    for i in range(len(str2)-1):
        if str2[i].isalpha() and str2[i+1].isalpha():
            setStr2.append(str2[i:i+2])

    setStr1.sort()
    setStr2.sort()

    intersection = 0
    union = 0

    p1 = 0
    p2 = 0

    while 1:
        if p1 == len(setStr1) and p2 == len(setStr2):
            break
        if p1 == len(setStr1):
            p2 += 1
            union += 1
            break
        elif p2 == len(setStr2):
            p1 += 1
            union += 1
            break
        if setStr1[p1] < setStr2[p2]:
            p1 += 1
            union += 1
            continue
        elif setStr1[p1] > setStr2[p2]:
            p2 += 1
            union += 1
            continue
        count1 = 1
        count2 = 1
        while p1 < len(setStr1)-1 and setStr1[p1] == setStr1[p1+1]:
            count1 += 1
            p1 += 1
        while p2 < len(setStr2)-1 and setStr2[p2] == setStr2[p2+1]:
            count2 += 1
            p2 += 1
        intersection += min(count1, count2)
        union += max(count1, count2)
        p1 += 1
        p2 += 1

    if union == 0:
        return 65536
    return int(intersection / union * 65536)

print(solution('FRANCE', 'french'))

File: jinny-l/season2/_230809/PGS_뉴스_클러스터링.java
package season2._230809;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

public class PGS_뉴스_클러스터링 {

    public int solution(String str1, String str2) {
        // 일괄 소문자 처리
        str1 = str1.toLowerCase();
        str2 = str2.toLowerCase();

        // 문자 2개씩 파싱
        Map<String, Integer> first = new HashMap<>();
        for (int i = 0; i < str1.length() - 1; i++) {
            String word = str1.substring(i, i + 2);

            if (!Pattern.matches("^[a-zA-Z]*$", word)) {
                continue;
            }

            first.put(word, first.getOrDefault(word, 0) + 1);
        }

        Map<String, Integer> second = new HashMap<>();
        for (int i = 0; i < str2.length() - 1; i++) {
            String word = str2.substring(i, i + 2);

            if (!Pattern.matches("^[a-zA-Z]*$", word)) {
                continue;
            }

            second.put(word, second.getOrDefault(word, 0) + 1);
        }

        if (first.isEmpty() && second.isEmpty()) {
            return 65536;
        }

        // (계산용) 중복 제거된 단어 구하기
        Set<String> words = new HashSet<>(){
            {
                addAll(first.keySet());
                addAll(second.keySet());
            }
        };

        // 교집합 구하기
        double intersection = 0;
        double union = 0;
        for (String word : words) {
            if (first.containsKey(word) && second.containsKey(word)) {
                intersection += Math.min(first.get(word), second.get(word));
                union += Math.max(first.get(word), second.get(word));

            } else {
                union += first.getOrDefault(word, 0);
                union += second.getOrDefault(word, 0);
            }
        }

        return (int) (intersection / union * (double) 65536);
    }
}
